sound/soc/sof/intel/hda.c

```diff
@@ -68,6 +68,8 @@ static int sdw_params_stream(struct device *dev,
 	config.type = SOF_DAI_INTEL_ALH;
 	config.dai_index = (link_id << 8) | (d->id);
 	config.alh.stream_id = alh_stream_id;
+	config.alh.rate = params_rate(params_data->hw_params);
+	config.alh.channels = params_channels(params_data->hw_params);
```

Comment:
@bardliao who configures the rate and channels present in the hw_params? it's not part of the topology definitions, so where does this come from? the front-end?
Say e.g. that you want to push 1ch on ALH1.2 and 1ch on ALH 2.2, how would this work?

I am not against providing the info to the firmware, but it seems a bit too magical for now. Thanks for details.

Reply:
> @bardliao who configures the rate and channels present in the hw_params? it's not part of the topology definitions, so where does this come from? the front-end?
> Say e.g. that you want to push 1ch on ALH1.2 and 1ch on ALH 2.2, how would this work?

It comes from soc core. The flow is soc_pcm_hw_params() -> soc_soc_dai_hw_params() ->intel_hw_params() -> intel_params_stream() -> sdw_params_stream(). 
All ALH dais on the same PCM will get the same hw_params. But, we can configure it to what we want in sdw_params_stream().

Reply:
@bardliao aren't you making big assumptions about what the demux does here and how it's configured? The SOC code  does not have any notion of demux/aggregration, so if the demux configuration was to pass 1 channel to ALH1.2 and 1 channel to ALH2.2, then what would the firmware do?

We also need to consider the capture case, if you have multiple devices that capture from 2 or more devices, then do you really want to send to each ALH DAI the configuration that comes from the SOC core?

I must be missing something here.

Reply:
> @bardliao aren't you making big assumptions about what the demux does here and how it's configured? The SOC code does not have any notion of demux/aggregration, so if the demux configuration was to pass 1 channel to ALH1.2 and 1 channel to ALH2.2, then what would the firmware do?
> 
> We also need to consider the capture case, if you have multiple devices that capture from 2 or more devices, then do you really want to send to each ALH DAI the configuration that comes from the SOC core?
> 
> I must be missing something here.

Let's config it on topology :)